lekko_client/__init__.py

```diff
@@ -1,9 +1,144 @@
 """Lekko Python SDK Client"""
 
+import functools
+from dataclasses import dataclass
+from threading import RLock
+from typing import Any, Dict, Optional, Type
+
+from google.protobuf.message import Message as ProtoMessage
+
+from lekko_client import exceptions
+from lekko_client.clients import (
+    CachedBackendClient,
+    CachedGitClient,
+    Client,
+    ConfigServiceClient,
+)
+from lekko_client.constants import LEKKO_API_URL, LEKKO_SIDECAR_URL  # noqa
+from lekko_client.stores import MemoryStore
+
 __version__ = "0.1.4"
 
-LEKKO_API_URL = "prod.api.lekko.dev:443"
-LEKKO_SIDECAR_URL = "localhost:50051"
+__client: Client
+__client_lock = RLock()
+
+
+@dataclass(kw_only=True)
+class Config:
+    owner_name: str
+    repo_name: str
+    api_key: Optional[str] = None
+    context: Optional[Dict[str, Any]] = None
+    lekko_uri: str = LEKKO_API_URL
+
+
+@dataclass(kw_only=True)
+class SidecarConfig(Config):
+    lekko_uri = LEKKO_SIDECAR_URL
+
+
+@dataclass(kw_only=True)
+class APIConfig(Config):
+    pass
+
+
+@dataclass(kw_only=True)
+class CachedServerConfig(Config):
+    pass
+
+
+@dataclass(kw_only=True)
+class CachedGitConfig(Config):
+    git_repo_path: str
+
+
+def initialize(config: Config) -> Client:
+    global __client
+    with __client_lock:
+        if __client:
+            __client.close()
+        match config:
+            case APIConfig(_) | SidecarConfig(_):
+                __client = ConfigServiceClient(
+                    config.lekko_uri, config.owner_name, config.repo_name, config.api_key, config.context
+                )
+            case CachedGitConfig(_):
+                __client = CachedGitClient(
+                    config.lekko_uri,
+                    config.owner_name,
+                    config.repo_name,
+                    MemoryStore(),
+                    config.git_repo_path,
+                    config.api_key,
+                    config.context,
+                )
+            case CachedServerConfig(_):
+                __client = CachedBackendClient(
+                    config.lekko_uri, config.owner_name, config.repo_name, MemoryStore(), config.api_key, config.context
+                )
+            case _:
+                raise exceptions.LekkoError("Unknown client mode")
+        return __client
+
+
+def set_client(client: Client):
+    global __client
+    with __client_lock:
+        if __client:
+            __client.close()
+        __client = client
+
+
+def __get_safe(func):
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        with __client_lock:
```

Comment:
I wonder what are the performance implications of getting a lock for every read (especially if this is non-cached client). AFAIK, python doesn't have rw locks in stdlib and we probably don't want to invent a bicycle here :)
What if we don't grab a lock on reads and declare that `initialize` and `set_client` should be used during initialization only and are not safe to call after the client is in use.